contracts/NonFungibleToken-v2.cdc

```diff
@@ -0,0 +1,158 @@
+that implement these interfaces to do various things with the tokens.
+
+By using resources and interfaces, users of NFT smart contracts can send
+and receive tokens peer-to-peer, without having to interact with a central ledger
+smart contract.
+
+To send an NFT to another user, a user would simply withdraw the NFT
+from their Collection, then call the deposit function on another user's
+Collection to complete the transfer.
+
+*/
+
+import MetadataViews from "./MetadataViews.cdc"
+
+/// The main NFT contract interface. Other NFT contracts will
+/// import and implement this interface
+///
+pub contract NonFungibleToken {
+
+    /// Interface that the NFTs have to conform to
+    ///
+    pub resource interface NFT { //: MetadataViews.Resolver {
+        /// The unique ID that each NFT has
+        pub let id: UInt64
+
+        pub fun getViews(): [Type]
+        pub fun resolveView(_ view: Type): AnyStruct?
+    }
+
+    /// Interface to mediate withdraws from the Collection
+    ///
+    pub resource interface Provider {
+        /// withdraw removes an NFT from the collection and moves it to the caller
+        pub fun withdraw(withdrawID: UInt64): @AnyResource{NFT} {
+            post {
+                result.id == withdrawID: "The ID of the withdrawn token must be the same as the requested ID"
+            }
+        }
+    }
+
+    /// Interface to mediate withdraws from the Collection
+    ///
+    pub resource interface Transferable {
+        /// withdraw removes an NFT from the collection and moves it to the caller
+        pub fun transfer(id: UInt64, receiver: Capability<&AnyResource{Receiver}>): @AnyResource{NFT}
+    }
+
+    /// Interface to mediate deposits to the Collection
+    ///
+    pub resource interface Receiver {
+
+        /// deposit takes an NFT as an argument and adds it to the Collection
+        ///
+        pub fun deposit(token: @AnyResource{NFT})
+
+        /// getAcceptedTypes returns a list of NFT types that this receiver accepts
+        pub fun getAcceptedTypes(): [Type]
+    }
+
+    /// Interface that an account would commonly 
+    /// publish for their collection
+    pub resource interface CollectionPublic { //: MetadataViews.ResolverCollection {
+        pub fun deposit(token: @AnyResource{NFT})
+        pub fun borrowViewResolver(id: UInt64): &{MetadataViews.Resolver}
+        pub fun getIDs(): [UInt64]
+        pub fun borrowNFT(id: UInt64): &AnyResource{NFT}
+    }
+
+    /// Requirement for the concrete resource type
+    /// to be declared in the implementing contract
+    ///
+    pub resource interface Collection { //: Provider, Receiver, Transferable, CollectionPublic, MetadataViews.ResolverCollection {
+
+        /// Paths for the collection
+        pub let storagePath: StoragePath
+        pub let publicPath: PublicPath
+        pub let privateProviderPath: PrivatePath
+
+        /// Dictionary to hold the NFTs in the Collection
```

Comment:
Maybe we can define something like another struct here: Something like Link : Type and Path , main problem in mainnet is linking currently, so maybe some improvement there would be nice. 

something like roughly:
```[
`Link(Path:/storage/TopShotCollection, Type: Type<TS.Collection>)`,
`Link(Path:/public/TopShotCollection, Type: Type<{TS.CollectionPublic>})`,
`Link(Path:/private/TopShotCollection, Type: Type<{NFT.Provider}>)`
]
```

Reply:
What's the anticipated usage for this? Might make more sense to make it a dictionary so that it's easy to access these by type. Otherwise transactions are looping over an array looking for the type they care about

Reply:
I'm changing the `getAcceptedTypes` return type to a dictionary, but I think @bluesign was talking about how to store path and link information actually. I'm not sure what the proposal here means though.

Reply:
I was suggesting instead of exposing path, maybe we can expose path + type information. So user of the contract, if setting new can now know, how to setup the account correctly.

Reply:
Isn't that what the `MetadataViews.NFTCollectionData` view is for?